pkg/operator/kubeletversionskewcontroller/kubelet_version_skew_controller.go

```diff
@@ -0,0 +1,258 @@
+	KubeletMinorVersionUnsupportedReason            = "KubeletMinorVersionUnsupported"
+	KubeletMinorVersionAheadReason                  = "KubeletMinorVersionAhead"
+)
+
+// KubeletVersionSkewController sets Upgradeable=False if the kubelet
+// version on a node prevents upgrading to a supported OCP version.
+//
+// If OCP minor version is odd, kubelet versions older than 1 minor
+// version of the API server version set Upgradeable=False.
+//
+// If OCP minor version is even, kubelet versions must be in sync with
+// API server version in order to set Upgradeable=True.
+type KubeletVersionSkewController interface {
+	factory.Controller
+}
+
+func NewKubeletVersionSkewController(
+	operatorClient v1helpers.OperatorClient,
+	kubeInformersForNamespaces v1helpers.KubeInformersForNamespaces,
+	recorder events.Recorder,
+) *kubeletVersionSkewController {
+	c := &kubeletVersionSkewController{
+		operatorClient:     operatorClient,
+		nodeLister:         kubeInformersForNamespaces.InformersFor("").Core().V1().Nodes().Lister(),
+		apiServerVersion:   semver.MustParse(status.VersionForOperandFromEnv()),
+		minUpgradeableSkew: minUpgradeableKubeletSkewForOCPVersion(status.VersionForOperatorFromEnv()),
+	}
+	c.Controller = factory.New().
+		WithSync(c.sync).
+		WithInformers(kubeInformersForNamespaces.InformersFor("").Core().V1().Nodes().Informer()).
+		ToController("KubeletVersionSkewController", recorder.WithComponentSuffix("kubelet-version-skew-controller"))
+	return c
+}
+
+func minUpgradeableKubeletSkewForOCPVersion(v string) int {
+	switch semver.MustParse(v).Minor % 2 {
+	case 0: // even OCP versions
+		return 0
+	case 1: // odd OCP versions
+		return -1
+	default:
+		panic("should not happen")
+	}
+}
+
+type kubeletVersionSkewController struct {
+	factory.Controller
+	operatorClient     v1helpers.OperatorClient
+	nodeLister         corev1listers.NodeLister
+	apiServerVersion   semver.Version
+	minUpgradeableSkew int
+}
+
+func (c *kubeletVersionSkewController) sync(_ context.Context, _ factory.SyncContext) error {
+	operatorSpec, _, _, err := c.operatorClient.GetOperatorState()
+	if err != nil {
+		return err
+	}
+	if !management.IsOperatorManaged(operatorSpec.ManagementState) {
+		return nil
+	}
+
+	nodes, err := c.nodeLister.List(labels.Everything())
+	if err != nil {
+		return err
+	}
+	sort.Sort(byName(nodes))
+
+	var errors nodeKubeletInfos
+	var skewedUnsupported nodeKubeletInfos
+	var skewedLimit nodeKubeletInfos
+	var skewedButOK nodeKubeletInfos
+	var synced nodeKubeletInfos
+	var unsupported nodeKubeletInfos
+
+	// for each node, check kubelet version
+	for _, node := range nodes {
+		kubeletVersion, err := nodeKubeletVersion(node)
+		if err != nil {
+			runtime.HandleError(fmt.Errorf("unable to determine kubelet version on node %s: %w", node.Name, err))
+			errors = append(errors, nodeKubeletInfo{node: node.Name, err: err})
+			continue
+		}
+		minSupportedSkew := c.minUpgradeableSkew - 1
```

Comment:
I don't think this is right.  Can we think about this in terms of what skew is supported, instead of what skew allows `Upgradeable=False`  or `Upgradeable=True`?  Then we can derive `Upgradeable` from support.  Constraints from Scott:

* If OCP minor version is odd, kubelet versions older than 1 minor version of the API server version set `Upgradeable=False`.
* If OCP minor version is even, kubelet versions must be in sync with API server version in order to set `Upgradeable=True`.

So 4.9 (odd) is `Upgradeable=True` with skew 0 and skew 1, which means 4.10 (even) supports skews 0 through 2.

And 4.8 (even) is `Upgradeable=True` with skew 0, which means 4.9 (odd) supports skew 0 and skew 1.

So for odd, `minUpgradeableSkew` is 1, but `minSupportedSkew` is also 1, right?
And for even, `minUpgradeableSkew` is 0 and `minSupportedSkew` is 2.
i.e. even is generous, but knows that odd will be stingy.  Odd is stingy, but knows even will be generous.

Reply:
I have re-worked the logic to work with a `minSupportedSkew` and `minSupportedSkewNextVersion` instead of a `minUpgradableSkew`.